src/wasm_udf/wasm.rs

```diff
@@ -11,22 +15,518 @@ use datafusion::{
+                    }),
+                },
+            )
+            .map(|a| Arc::new(a) as ArrayRef)
+        }
+        CreateFunctionDataType::I64 | CreateFunctionDataType::BIGINT => {
+            decode_udf_result_primitive_array::<arrow::datatypes::Int64Type>(
+                encoded_results,
+                &|v| {
+                    v.as_i64().ok_or(DataFusionError::Internal(format!(
+                        "Expected to find i64 value, but received {:?} instead",
+                        v
+                    )))
+                },
+            )
+            .map(|a| Arc::new(a) as ArrayRef)
+        }
+        CreateFunctionDataType::CHAR
+        | CreateFunctionDataType::VARCHAR
+        | CreateFunctionDataType::TEXT => {
+            let mut decoded_results = Vec::with_capacity(encoded_results.len());
+            for i in encoded_results {
+                decoded_results.push(Some(i.as_str().ok_or(
+                    DataFusionError::Internal(format!(
+                        "Expected to find string value, received {:?} instead",
+                        &i
+                    )),
+                )?));
+            }
+            Ok(Arc::new(decoded_results.iter().collect::<StringArray>()) as ArrayRef)
+        }
+        CreateFunctionDataType::DATE => decode_udf_result_primitive_array::<
+            arrow::datatypes::Date32Type,
+        >(encoded_results, &|v| match v
+            .as_i64()
+            .ok_or(DataFusionError::Internal(format!(
+                "Expected to find i64 value, but received {:?} instead",
+                v
+            ))) {
+            Err(e) => Err(e),
+            Ok(v_i64) => i32::try_from(v_i64).map_err(|e| {
+                DataFusionError::Internal(format!(
+                    "Error converting i64 to i32 (for date): {:?}",
+                    e
+                ))
+            }),
+        })
+        .map(|a| Arc::new(a) as ArrayRef),
+        CreateFunctionDataType::TIMESTAMP => decode_udf_result_primitive_array::<
+            arrow::datatypes::TimestampNanosecondType,
+        >(encoded_results, &|v| {
+            v.as_i64().ok_or(DataFusionError::Internal(format!(
+                "Expected to find i64 value, but received {:?} instead",
+                v
+            )))
+        })
+        .map(|a| Arc::new(a) as ArrayRef),
+        CreateFunctionDataType::BOOLEAN => {
+            let mut decoded_results = Vec::with_capacity(encoded_results.len());
+            for i in encoded_results {
+                decoded_results.push(Some(i.as_bool().ok_or(
+                    DataFusionError::Internal(format!(
+                        "Expected to find bool value, received {:?} instead",
+                        &i
+                    )),
+                )?));
+            }
+            Ok(Arc::new(
+                decoded_results
+                    .iter()
+                    .collect::<arrow::array::BooleanArray>(),
+            ) as ArrayRef)
```

Comment:
similar to above (assuming that `Some(...)` wrapper is redundant):

```suggestion
            let decoded_results = encoded_results.iter().map(|i| (i.as_bool().ok_or(
                DataFusionError::Internal(format!(
                   "Expected to find bool value, received {:?} instead",
                    &i
                )),
            ))).collect::<Result<BooleanArray>>()?;
            Ok(Arc::new(decoded_results) as ArrayRef)
```

Reply:
(the Some is required as we discovered but it can be re-added inside of the closure)